src/internal_rep.rs

```diff
@@ -1338,6 +1348,277 @@ fn call_to_fc_rules<'a>(
+                name: CascadeString::from("fscontext_type"),
+                default: None,
+            },
+            types,
+            None,
+        )?,
+        FunctionArgument::new(
+            &DeclaredArgument {
+                param_type: CascadeString::from("path"),
+                is_list_param: false,
+                name: CascadeString::from("path_regex"),
+                default: Some(Argument::Quote(CascadeString::from("\"/\""))),
+            },
+            types,
+            None,
+        )?,
+        FunctionArgument::new(
+            &DeclaredArgument {
+                param_type: CascadeString::from("obj_class"), //TODO: not really
+                is_list_param: true,
+                name: CascadeString::from("file_type"),
+                default: Some(Argument::List(vec![])),
+            },
+            types,
+            None,
+        )?,
+    ];
+    let validated_args = validate_arguments(c, &target_args, types, class_perms, context, file)?;
+    let mut args_iter = validated_args.iter();
+    let mut ret = Vec::new();
+
+    let context_str = args_iter
+        .next()
+        .ok_or_else(|| ErrorItem::Internal(InternalError::new()))?
+        .get_name_or_string(context)?;
+    let fs_context = match Context::try_from(context_str.to_string()) {
+        Ok(c) => c,
+        Err(_) => {
+            return Err(CascadeErrors::from(ErrorItem::Compile(CompileError::new(
+                "Invalid context",
+                file,
+                context_str.get_range(),
+                "Cannot parse this into a context",
+            ))))
+        }
+    };
+    let fs_name = args_iter
+        .next()
+        .ok_or_else(|| ErrorItem::Internal(InternalError::new()))?
+        .get_name_or_string(context)?
+        .to_string();
+    let fscontext_str = args_iter
+        .next()
+        .ok_or_else(|| ErrorItem::Internal(InternalError::new()))?
+        .get_name_or_string(context)?;
+    let fscontext_type = match fscontext_str.to_string().parse::<FSContextType>() {
+        Ok(f) => f,
+        Err(_) => {
+            return Err(CascadeErrors::from(ErrorItem::Compile(CompileError::new(
+                "Not a valid file system type",
+                file,
+                fscontext_str.get_range(),
+                "File system type must be 'xattr', 'task', 'trans', or 'genfscon'",
+            ))));
+        }
+    };
+    let regex_string_arg = args_iter
+        .next()
+        .ok_or_else(|| ErrorItem::Internal(InternalError::new()))?;
+    let regex_string = regex_string_arg.get_name_or_string(context)?.to_string();
+    let file_types_arg = args_iter
+        .next()
+        .ok_or_else(|| ErrorItem::Internal(InternalError::new()))?;
+    let file_types = file_types_arg.get_list(context)?;
+
+    match fscontext_type {
+        FSContextType::XAttr | FSContextType::Task | FSContextType::Trans => {
+            // The 'regex_string_arg.get_range().is_none()' is a hacky way to
```

Comment:
Eww.  I'm not immediately sure what a better answer is though.  This is mostly a note to self to think about this more.  I'm worried that overloading the usage of the range like this will cause us problems down the road, since we're tying this user to the implementation detail.  I can't think of any reason it wouldn't work in the current code base though...

Reply:
Probably the long term answer will be to augment the TypeInstance with information about where this instance came from.  That seems like a longer term answer.  I think this way is fine for now.